src/clj_druid/client.clj

```diff
@@ -1,92 +1,72 @@
             [clj-http.client :as http]))
 
-(def default-timeout 5000)
-
-(def nodes-list
-  "contains all zk nodes discovered"
-  (atom []))
-
-(defn reset-node-list
-  "update node list atom"
-  [nodes]
-  (reset! nodes-list nodes))
-
-(defn make-node-path
-  "make a zk path string"
-  [discovery-path node-type]
-
-  (str discovery-path "/" node-type))
+(def service-fn
+  "Service discovery function that gives a node list from zk
+  or a user defined host."
+  (atom nil))
 
 (defn make-host-http-str
-  "make an http url string from a zk node entry"
-  [c]
-
+  "make an http url string from a curator intance"
+  [instance]
   (str "http://"
-       (get c "address")
+       (.getAddress instance)
        ":"
-       (get c "port")
+       (.getPort instance)
        "/druid/v2/"))
 
-(defn zk-watch-node-list
-  "Retrieve hosts from zk discovery"
-  [zk-client path]
-
-  (-<>> path
-        (zk/children zk-client <>
-                     :watch? true
-                     :watcher #(do (println %)
-                                   (zk-watch-node-list zk-client path)))
-
-        (map #(data/to-string (:data (zk/data zk-client (str  path "/" %)))))
-        (map #(json/read-str %))
-        (map #(make-host-http-str %))
-        (vec)
-        (reset-node-list)))
-
-(defn from-zookeeper
+(defn zookeeper-service-fn
   "Maintain a druid http server list from zookeeper"
   [config]
-
-  (let [node-path (make-node-path (:discovery-path config)
-                                  (:node-type config))
-
-        zk-client (zk/connect (:host config))]
-
-    (zk-watch-node-list zk-client node-path)))
-
-(defn from-user
+  (fn []
+    (let [client (curator-framework (:host config))
+          _ (.start client)
+          discovery (service-discovery client nil :base-path (:discovery-path config))
+          _ (.start discovery)
+          sc (service-cache discovery (:node-type config))
+          _ (.start sc)]
+      (map make-host-http-str (instances sc)))))
+
+(defn user-service-fn
   "Maintain a druid http server list from user"
   [hosts]
-  (reset-node-list hosts))
+  (constantly hosts))
+
+(defn nodes-list
+  "Provide a node list from zk or a user defined host"
+  []
+  (when-not @service-fn
+    (throw (IllegalStateException. "The druid client is not initialized.")))
+  (let [nodes (@service-fn)]
+    (if (empty? nodes)
+      (throw (Exception.
+              "No druid node available for query"))
+      nodes)))
 
 (defn randomized
   "Take a random host"
   []
-  (if (empty? @nodes-list)
-    (throw (Exception.
-            "No druid node available for query")))
-
-  (rand-nth @nodes-list))
+  (rand-nth (nodes-list)))
 
 (defn fixed
   "Always take first host"
   []
-  (first @nodes-list))
+  (first (nodes-list)))
 
 (defn connect
   "Create a druid client from zk or
   a user defined host"
   [params]
-  (if (:zk params)
-    (from-zookeeper (:zk params))
-    (from-user (:hosts params))))
+  (reset! service-fn
```

Comment:
I would rather not use a shared atom here, what would happen if you need to connect to different druid instances?